specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,467 @@
+1. If `CursorIsValid(Cursor)` is `false`, return `null` (or throw an error)
+2. Get the `link` from `cursor.node.links` at `cursor.index`
+3. return the `link`
+
+### Get(ProllyNode, key): value
+
+1. Get a `cursor` from `CursorAtItem(ProllyNode, key)`
+2. If `CursorIsValid(cursor)` is `false`, return an error (key not found)
+3. Get `currentKey` from `CursorGetKey(Cursor)`
+4. If `currentKey` is not bytewise equal to `key`, return an error (key not found)
+5. Get `value` from `CursorGetValue(Cursor)`
+6. Return `value`
+
+### weibullCDF(x, K Int, L Int) : int
+
+Note: CDF (Cumulative Distribution Function) of the ][Weibull probability distribution](https://en.wikipedia.org/wiki/Weibull_distribution)
+ - return `-exp(-pow(x/L),K)`
+
+### WeibullThreshold(ProllyNode, key, value) : Boolean
+
+TODO: cleanup
+
+WeibullThreshold returns true if we should split at |hash| for a given record inserted into a chunk of Size |Size|, where the record's Size is |thisSize|. |Size| is the Size of the chunk after the record is inserted, so includes |thisSize| in it.
+
+WeibullThreshold attempts to form chunks whose sizes match the weibull distribution.
+
+The logic is as follows: given that we haven't split on any of the records up to |Size - thisSize|,  the probability that we should split on this record is (CDF(end) - CDF(start)) / (1 - CDF(start)), or, the precentage of the remaining portion of the CDF that this record actually covers. We split is |hash|, treated as a uniform random number between [0,1), is less than this percentage.
+
+0. Get the `{bytes, cid}` from `saveNode(value)`
+1. `hash = hash(key + saveNode(value).cid)` *TODO: hash needs to be well-defined*
+2. `itemSize = len(key) + len(value)`
+3. Set `prevItemsSize` to the sum of the length of all the node's previous items (excluding the current one)
+4. `start = weibullCDF(prevItemsSize, K, L)`
+5. `end = weibullCDF(prevItemsSize + itemSize, K, L)`
+6. `p = float(hash)/maxUInt32`
+7. `d = 1 - start`
+8. If `d <= 0`: return `true` (this should realistically never occour?)
+9. `target = (end - start)/d`
+10. return `p < target`
+
+### HashByteArray([]byte, offset) : Boolean
+
+*TODO: Description*
+*TODO: current implementation applies the min/maxChunkSize to offset. Do we want this?*
+- For each `byte`:
+    1. Increment `offset` by 1
+    2. Feed `byte` to rolling hash *TODO Details for hashing, salting*
+    3. Set `hash` to hashSum of rolling hash
+    4. `pattern = (1<<(15 - (offset >> 10))) - 1`
+    5. If `hash&pattern == pattern` return `true`
+- Return `false`
+
+### RollingHash(ProllyNode, index) : Boolean
+
+*TODO: Rolling hash window*
+1. Set `offset` to the number of bytes of the elements of previous keys and values
+2. If `HashByteArray(key) == true`: return `true`
+3. Add number of bytes of key to `offset`
+4. If `HashByteArray() == true`: return `true`
+5. return `false`
+
+### ShouldCreateBoundry(ProllyNode, key) : Boolean
```

Comment:
I think we need a structure `Splitter`(ChunkSplitter) with state(such as offset, hash in RollingHash). We put k1v1, k2v2... into Splitter one by one and change the state then return whether boundary generates, if true we need reset the Splitter state and save the generated node.
We need a function to receive the input k/v one by one like append(k, v []byte). It must input in order, can't skip, repeat or draw back.
If only this function, it miss important information is where the computing starts. i.e, which range is the boundary generated for? Because I only see a pair of k/v is used in computing

Reply:
Maybe we can have different code paths for tracking state based on the chunking strategy? Or have another way to represent what state the chunking strategy is holding?